src/animation.rs

```diff
@@ -1,20 +1,40 @@
+use std::iter::Iterator;
+
 use num_rational::Ratio;
 
 use buffer::RgbaImage;
+use image::ImageResult;
 
-/// Holds the frames of the animated image
-pub struct Frames {
-    frames: Vec<Frame>,
-    current_frame: usize,
+/// An implementation dependent iterator, reading the frames as requested
+pub struct Frames<'a> {
+    iterator: Box<Iterator<Item = ImageResult<Frame>> + 'a>
 }
 
-impl Frames {
-    /// Contructs a new frame iterator
-    pub fn new(frames: Vec<Frame>) -> Frames {
-        Frames {
-            frames,
-            current_frame: 0,
+impl<'a> Frames<'a> {
+    /// Creates a new `Frames` from an implementation specific iterator.
+    pub fn new(iterator: Box<Iterator<Item = ImageResult<Frame>> + 'a>) -> Self {
+        Frames { iterator }
+    }
+
+    /// Steps through the iterator from the current frame until the end and pushes each frame into
+    /// a `Vec`.
+    /// If en error is encountered that error is returned instead.
+    pub fn collect_frames(mut self) -> ImageResult<Vec<Frame>> {
+        let mut result = Vec::new();
```

Comment:
This whole procedure should be equivalent to `self.collect::<ImageResult<Vec<Frame>>()` if I'm not mistaken. The `FromIterator` impls for `Result` is very broad.